src/AFQMC/Estimators/newBackPropagatedEstimator.hpp

```diff
@@ -0,0 +1,294 @@
+    wfn0.getReferencesForBackPropagation(Refs_[0]);
+    for(int iw=1; iw<wset.size(); ++iw)
+      for(int ref=0; ref<nrefs; ++ref)
+       copy_n(Refs_[0][ref].origin()+n0,n1-n0,Refs_[iw][ref].origin()+n0);
+    TG.TG_local().barrier();
+
+    //3. propagate backwards the references
+    prop0.BackPropagate(bp_step,nStabalize,wset,Refs_,detR);
+
+    //4. calculate properties, only rdm now but make a list or properties later
+    // adjust weights here is path restoration
+    stdCVector wgt(iextensions<1u>{wset.size()});
+    wset.getProperty(WEIGHT,wgt);
+    if(path_restoration) {
+      auto&& factors(wset.getWeightFactors());
+      for(int k=0; k<bp_step; k++)  
+        for(int i=0; i<wgt.size(); i++) 
+          wgt[i] *= factors[k][i];
+    } else if(!importanceSampling) {
+      stdCVector phase(iextensions<1u>{wset.size()});
+      wset.getProperty(PHASE,phase);
+      for(int i=0; i<wgt.size(); i++) wgt[i] *= phase[i];
+    }
+    observ0.accumulate(iav, wset, Refs_, wgt, detR, importanceSampling);
+
+    if(bp_step == max_nback_prop) {
+      // 5. setup for next block 
+      for(auto it=wset.begin(); it<wset.end(); ++it)
+        it->setSlaterMatrixN(); 
+      wset.setBPPos(0);
+
+      // 6. increase block counter
+      iblock++;
+      accumulated_in_last_block=true;
+    }
+    AFQMCTimers[back_propagate_timer]->stop();
+  }
+
+  void tags(std::ofstream& out) {
+    if(writer) 
+      out<<"BP_timer ";
+  }
+
+  void print(std::ofstream& out, hdf_archive& dump, WalkerSet& wset)
+  {
+    // I doubt we will ever collect a billion blocks of data.
+    int n_zero = 9;
+    if(writer) {
+      out<<std::setprecision(5)
+                    <<AFQMCTimers[back_propagate_timer]->get_total() <<" ";
+      AFQMCTimers[back_propagate_timer]->reset();  
+    }
+    if(accumulated_in_last_block) {
+      if(writer && first) { 
+        first=false;
+        int nave(nback_prop_steps.size());
+        int nref(detR.size(1));
+        if(write_metadata) {
+          dump.push("Observables");
+          dump.push("BackPropagated");
+          dump.push("Metadata");
+          dump.write(nback_prop_steps, "BackPropSteps");
+          dump.write(nave, "NumAverages");
+          dump.write(nref, "NumReferences");
+          dump.pop();
+          dump.pop();
+          dump.pop();
+          write_metadata = false;
+        }
+      }
+      if(writer) { 
+        dump.push("Observables");
+        dump.push("BackPropagated");
+      }
+      observ0.print(iblock,dump);
+      if(writer) { 
+        dump.pop();
+        dump.pop();
+      }  
+    }
+  }
+
+  private:
+
+  TaskGroup_& TG;
+
+  WALKER_TYPES walker_type;
+
+  bool writer;
+  bool accumulated_in_last_block;
+
+  int max_nback_prop;
+  std::vector<int> nback_prop_steps;  
+
+  FullObsHandler observ0; 
+
+  Wavefunction& wfn0;
+
+  Propagator& prop0;
+
+  // The first element of data stores the denominator of the estimator (i.e., the total
```

Comment:
This comment needs to be deleted at some point.

Reply:
Done